apis/datadoghq/v1alpha1/datadogdashboard_validation.go

```diff
@@ -0,0 +1,40 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package v1alpha1
+
+import (
+	"fmt"
+
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadogV1"
+	utilserrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+// IsValidDatadogDashboard use to check if a DatadogDashboardSpec is valid by checking
+// that the required fields are defined
+func IsValidDatadogDashboard(spec *DatadogDashboardSpec) error {
+	var errs []error
+	if spec.Title == "" {
+		errs = append(errs, fmt.Errorf("spec.Title must be defined"))
+	}
+
+	if spec.LayoutType == "" {
+		errs = append(errs, fmt.Errorf("spec.LayoutType must be defined"))
+	}
+
+	if spec.LayoutType != "" && !spec.LayoutType.isValid() {
+		errs = append(errs, fmt.Errorf("spec.LayoutType must be one of the values: %s or %s", DashboardLayoutTypeFree, DashboardLayoutTypeOrdered))
+	}
+
+	if spec.LayoutType != DashboardLayoutTypeOrdered && spec.ReflowType != nil {
+		errs = append(errs, fmt.Errorf("spec.ReflowType should only be set if layout type is 'ordered'"))
+	}
+
+	if spec.ReflowType != nil && !spec.ReflowType.IsValid() {
+		errs = append(errs, fmt.Errorf("spec.ReflowType must be one of the values: %s or %s", datadogV1.DASHBOARDREFLOWTYPE_AUTO, datadogV1.DASHBOARDREFLOWTYPE_FIXED))
```

Comment:
I have some reservation about validating the input *values* to the API. I checked and saw similar pattern in SLOs and I supposed this is following same pattern. Validating required fields not being null should be sufficient.

One problem with this approach is that when we release Operator it will be compile with these constants (reflow type, layout type) so if Dashboards and SLOs start supporting new types we will have to make changes here. Moreover, if types aren't correct, or combination isn't correct we are generating custom error instead of reporting API error which imo, is easier to investigate.